src/celengine/textlayout.cpp

```diff
@@ -0,0 +1,252 @@
+                began = false;
+            }
+            else
+            {
+                // bind the font and set the same info
+                font->bind();
+                font->setMVPMatrices(projection, modelview);
+            }
+        }
+    }
+}
+
+void TextLayout::setHorizontalAlignment(HorizontalAlignment value)
+{
+    if (horizontalAlignment != value)
+    {
+        if (began)
+            flushInternal(false);
+        horizontalAlignment = value;
+    }
+}
+
+void TextLayout::setScreenDpi(int value)
+{
+    auto floatValue = static_cast<float>(value);
+    if (screenDpi != floatValue)
+    {
+        if (began)
+            flushInternal(false);
+        screenDpi = floatValue;
+    }
+}
+
+void TextLayout::moveAbsolute(float newPositionX, float newPositionY, bool updateAlignment)
+{
+    if (positionX != newPositionX || positionY != newPositionY)
+    {
+        if (began)
+            flushInternal(false);
+        positionX = newPositionX;
+        positionY = newPositionY;
+        if (updateAlignment)
+            alignmentEdgeX = positionX;
+    }
+}
+
+void TextLayout::moveRelative(float dx, float dy, Unit unit, bool updateAlignment)
+{
+    float resolvedDx = getPixelSize(dx, unit);
+    float resolvedDy = getPixelSize(dy, unit);
+    if (resolvedDx != 0.0f || resolvedDy != 0.0f)
+    {
+        if (began)
+            flushInternal(false);
+        positionX += resolvedDx;
+        positionY += resolvedDy;
+        if (updateAlignment)
+            alignmentEdgeX = positionX;
+    }
+}
+
+void TextLayout::begin(const Eigen::Matrix4f &p, const Eigen::Matrix4f &m)
+{
+    if (font == nullptr) return;
+
+    // if already began, do not call bind
+    if (!began)
+        font->bind();
+    font->setMVPMatrices(p, m);
+    began = true;
+    projection = p;
+    modelview = m;
+}
+
+void TextLayout::render(std::string_view text)
+{
+    if (!began)
+        return;
+
+    std::vector<std::wstring> lines;
+    if (!processString(text, lines))
+        return;
+
+    bool firstLine = true;
+    for (const auto& line : lines)
+    {
+        if (firstLine)
+        {
+            firstLine = false;
+        }
+        else
+        {
+            // Reset to line start, and go to the next line
+            positionX = alignmentEdgeX;
+            positionY -= static_cast<float>(font->getHeight());
+        }
+        if (!line.empty())
+            renderLine(line);
+    }
+}
+
+void TextLayout::flush()
+{
+    flushInternal(true);
+}
+
+void TextLayout::end()
+{
+    if (!began)
+        return;
+
+    flushInternal(true);
+    if (font != nullptr)
+        font->unbind();
+    began = false;
+}
+
+std::pair<float, float> TextLayout::getCurrentPosition() const
+{
+    return {positionX, positionY};
+}
+
+int TextLayout::getLineHeight() const
+{
+    return (font == nullptr) ? 0 : font->getHeight();
+}
+
+int TextLayout::getTextWidth(std::string_view text, const std::shared_ptr<TextureFont> &font)
```

Comment:
This should take `const TextureFont*` rather than `const std::shared_ptr<TextureFont>&` since it isn't doing anything with ownership of the font.

Reply:
fixed